docs: clearify backend support for setting modtime only

albertony opened the pull request:
#### What is the purpose of this change?

It is confusing for users to know what "ModTime" support in a backend really means.

This change documents backends that support storing and reading modtimes (`ModTime`), at least some kind of usable timestamp, but does not support changing the modtime-only (`SetModTime`). Marking these with `Partial` in the `ModTime` column features table. 

Existing version seemed to have a mix of `Yes` and `No` for backends supporting `ModTime` but not `SetModTime` (I guess it is confusing for developers too - and I'm including me here).

The following are set to Partial, because `ModTime` is implemented, but `SetModTime` returns `ErrorCantSetModTime` or `ErrorCantSetModTimeWithoutDelete`:
```
1Fichier: No -> Partial
Amazon Drive: No -> Partial
Dropbox: Yes -> Partial
FTP: No -> Partial
Google Photos: No -> Partial
HTTP: No -> Partial
Koofr: No -> Partial
Mega: No -> Partial
pCloud: Yes -> Partial
premiumize.me: No -> Partial
Seafile: No -> Partial
Sugarsync: No -> Partial
Tardigrade: Yes -> Partial
WebDAV: Yes -> Partial **Note: Supports modtimes when used with Owncloud and Nextcloud only**
Zoho WorkDrive: No -> Partial
```

Other changes:
```
QingStor: No -> Yes **Note: SetModTime is unsupported for objects bigger than 5GiB**
```

The only one left with a big fat `No` is Uptobox. Its implementation of ModTime just returns `time.Now()`.
- If it hadn't been for this we could have renamed the column to "SetModTime" and only used Yes/No values.

I removed the following sentence from docs, couldn't quite grasp what this meant in relation to the other things described:

> All cloud storage systems support some kind of date on the object and
> these will be set when transferring from the cloud storage system.

#### Was the change discussed in an issue or in the forum before?

https://github.com/rclone/rclone/issues/1066

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [ ] ~I have added tests for all changes in this PR if appropriate.~
- [X] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

ncw reviewed (commented):
Nice idea - thank you :-)

I think `Read Only` would be clearer than partial, though maybe we should use `-` / `R` / `R/W` like the `MimeType` column does?

What do you think?

albertony commented:
> maybe we should use `-` / `R` / `R/W` like the `MimeType` column does?

I like that idea! Will look into it.

ncw commented:
Just thinking this through a bit more

I think these are the categories...

1. `R/W`: read and write the modtime
    - backends like this return `Precision` <= time.Second and implement `SetModTime` 
    - drive / s3 / etc 
2. `R` / set the modtime on upload - but not generally with SetModTime
    - backends like this return `Precision` <= time.Second and `SetModTime` returns `ErrorCantSetModTime` or `ErrorCantSetModTimeWithoutDelete`
    - eg dopbox
3. `-` / return the upload time (or some other time eg Now, or the time the photo was taken) as the ModTime
    - these backends return `ModTimeNotSupported` from `Precision` and don't implement `SetModTime`
    - eg googlephotos / mega

I was thinking about splitting the categories in 3, because returning the upload time means you can do a sync with `--update` but if the time isn't the upload time (eg the current time or photo time) then that won't work.

We don't have a feature flag for ModTimeIsUseful but maybe we should!

albertony commented:
| Name                        | Precision                          | SetModTime                          | ModTime | Result
| --------------------------- |----------------------------------- | ----------------------------------- | ------- | ----- |
| 1Fichier                    | ModTimeNotSupported                | ErrorCantSetModTime                 |         | R     |
| Amazon Drive                | ModTimeNotSupported                | ErrorCantSetModTime                 |         | R     |
| Amazon S3                   | Nanosecond                         | OK                                  |         | R/W   |
| Backblaze B2                | Nanosecond                         | OK                                  |         | R/W   |
| Box                         | Second                             | OK                                  |         | R/W   |
| Citrix ShareFile            | Second                             | OK                                  |         | R/W   |
| Dropbox                     | Second                             | ErrorCantSetModTimeWithoutDelete    |         | R     |
| Enterprise File Fabric      | Unknown (returned by the API)      | OK                                  |         | R/W   |
| FTP                         | ModTimeNotSupported                | nil                                 |         | R     |
| Google Cloud Storage        | Nanosecond                         | OK                                  |         | R/W   |
| Google Drive                | Nanosecond                         | OK                                  |         | R/W   |
| Google Photos               | ModTimeNotSupported                | ErrorCantSetModTime                 |         | R     |
| HDFS                        | Second                             | OK                                  |         | R/W   |
| HTTP                        | Second                             | "error: http remotes are read only" |         | R     | **Note:** Precision is the remote http file system's modtime precision, which we have no way of knowing. We estimate at 1s.
| Hubic                       | Nanosecond                         | OK                                  |         | R/W   |
| Jottacloud                  | Second                             | PR5634                              |         | R/W   |
| Koofr                       | ModTimeNotSupported or ms (config) | ErrorCantSetModTimeWithoutDelete    |         | R     |
| Mail.ru Cloud               | Second                             | OK                                  |         | R/W   |
| Mega                        | ModTimeNotSupported                | ErrorCantSetModTime                 |         | R     |
| Memory                      | Nanosecond                         | OK                                  |         | R/W   |
| Microsoft Azure Blob Storage| Nanosecond                         | OK                                  |         | R/W   |
| OneDrive                    | Second                             | OK                                  |         | R/W   |
| OpenDrive                   | Second                             | OK                                  |         | R/W   |
| OpenStack Swift             | Nanosecond                         | OK                                  |         | R/W   |
| pCloud                      | Second                             | ErrorCantSetModTime                 |         | R     |
| premiumize.me               | ModTimeNotSupported                | ErrorCantSetModTime                 |         | R     |
| put.io                      | Second                             | OK                                  |         | R/W   |
| QingStor                    | ModTimeNotSupported                | OK but max 5GiB                     |         | R     |
| Seafile                     | ModTimeNotSupported                | ErrorCantSetModTime                 |         | R     |
| SFTP                        | Second                             | OK                                  |         | R/W   |
| SugarSync                   | ModTimeNotSupported                | ErrorCantSetModTime                 |         | R     |
| Tardigrade                  | Nanosecond                         | ErrorCantSetModTime                 |         | R     |
| Uptobox                     | ModTimeNotSupported                | ErrorCantSetModTime                 | Now     | -     |
| WebDAV                      | From API                           | ErrorCantSetModTime                 |         | R     | **Note:** Supports modtimes when used with Owncloud and Nextcloud only
| Yandex Disk                 | Nanosecond                         | OK                                  |         | R/W   |
| Zoho WorkDrive              | ModTimeNotSupported                | ErrorCantSetModTime                 |         | R     |
| Local                       | Probe disk                         | OK                                  |         | R/W   |

TODO:
- Tardigrade returns `Precision: Nanosecond` but no `SetModTime` support!
- QingStor returns `Precision: ModTimeNotSupported`, but supports `SetModTime` - for files under 5 GiB!
- Jottacloud: Assumes PR 5634 will be merged!
- Google Photos and Mega seem to return `ModTime`, so is not a plain `R`, not `-` as you indicated?
- FTP always returns `SetModTime: nil`, why not `ErrorCantSetModTime`?

ncw commented:
This is going to miss 1.57 alas, but we can point release it!

Very nice table :-)

> * Tardigrade returns `Precision: Nanosecond` but no `SetModTime` support!

That is OK, it just means that SetModTime will trigger a re-upload.

https://github.com/rclone/rclone/blob/c9d7248d8548ea43d6421f8eec555f2af6f28e66/fs/operations/operations.go#L242-L245

> * QingStor returns `Precision: ModTimeNotSupported`, but supports `SetModTime` - for files under 5 GiB!

Bleh!

> * Jottacloud: Assumes PR 5634 will be merged!

It is merged now :-)

> * Google Photos and Mega seem to return `ModTime`, so is not a plain `R`, not `-` as you indicated?

All the backends will return a `ModTime`, just not necessarily the one you hoped they would! For Google photos the mod time is the date the photo was taken or the time it was uploaded if gphotos can't read the EXIF data. Which is great for looking at your photos, but hopeless for syncing.

For a backend to be in class `R` the object has to have the modification time of when we uploaded it and a useful Precision.

So I think any backend in the table above which has Precision ModTimeNotSupported needs to be `-`

> * FTP always returns `SetModTime: nil`, why not `ErrorCantSetModTime`?

I think that is a bug!

I wonder whether it would be worth writing an `rclone test modtime` which uploads a file checks to see what timestamp it actually got and tries resetting the timestamp to see what happens?

ncw commented on docs/content/overview.md:
```diff
@@ -94,13 +96,26 @@ systems they must support a common hash type.
+modification time only on a files in a `mount` will be silently ignored.
+
+Storage systems with `R/W`, for read/write, in the ModTime column, means
+they do support modtime-only operations. An `-` means it does not
```
This should probably say something like

> A `-` means the modification read on objects is not the modification time of the file when uploaded. It is most likely the time the file was uploaded, or possibly something else (like the time the picture was taken in Google Photos).

albertony replied:
Nice! I incorporated that description in a new version.

ncw commented on docs/content/overview.md:
```diff
@@ -16,44 +16,44 @@ Here is an overview of the major features of each cloud storage system.
-| Yandex Disk                  | MD5         | Yes     | No               | No              | R         |
-| Zoho WorkDrive               | -           | No      | No               | No              | -         |
-| The local filesystem         | All         | Yes     | Depends          | No              | -         |
+| 1Fichier                     | Whirlpool   | R       | No               | Yes             | R         |
```
I think all the ones which has ModTime `No` here need to be `-` rather than `R`

-  amazonclouddrive
-  fichier
-  ftp (caveats apply - for a fully supported FTP server this should be `R/W` not `-`)
-  googlephotos
-  koofr
-  mega
-  premiumizeme
-  qingstor
-  seafile
-  sia
-  sugarsync
-  uptobox
-  webdav (caveats apply)
-  zoho

albertony replied:
Yes! `ModTime: No`, or in new version `ModTime: -`, in that table directed at end users, are not really related to the `ModTime` function in code but more to the `Precision` function and corresponds to it returning `ModTimeNotSupported`!

ncw reviewed (commented):
I had a look through the table...

We should probably add an integration test that ensures that 

if Precision == NotSupported then check SetModTime returns Error - I think that should always be an error?

albertony commented:
Thanks for the explanations!
I take it Precision is the important property for syncing, not ModTime which I originally based the classification on...

albertony commented:
Pushed a new version of the docs, and below is an update to the more detailed table. More to dig into here, but 

<details>
<summary>Updated table: Click to expand!</summary>

| Name                        | Precision                          | SetModTime                          | ModTime | Result| Comment                                                                                                           |
| --------------------------- |----------------------------------- | ----------------------------------- | ------- | ----- | ----------------------------------------------------------------------------------------------------------------- |
| 1Fichier                    | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Amazon Drive                | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Amazon S3                   | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Backblaze B2                | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Box                         | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Citrix ShareFile            | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Dropbox                     | Second                             | ErrorCantSetModTimeWithoutDelete    |         | R     |                                                                                                                   |
| Enterprise File Fabric      | Unknown (returned by the API)      | OK                                  |         | R/W   |                                                                                                                   |
| FTP                         | ModTimeNotSupported                | nil                                 |         | -     | Bug, SetModTime should have returned ErrorCantSetModTime                                                          |
| Google Cloud Storage        | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Google Drive                | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Google Photos               | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| HDFS                        | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| HTTP                        | Second                             | "error: http remotes are read only" |         | R     | Precision is the remote http file system's modtime precision, which we have no way of knowing. We estimate at 1s. |
| Hubic                       | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Jottacloud                  | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Koofr                       | ModTimeNotSupported or ms (config) | ErrorCantSetModTimeWithoutDelete    |         | -     |                                                                                                                   |
| Mail.ru Cloud               | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Mega                        | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Memory                      | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Microsoft Azure Blob Storage| Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| OneDrive                    | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| OpenDrive                   | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| OpenStack Swift             | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| pCloud                      | Second                             | ErrorCantSetModTime                 |         | R     |                                                                                                                   |
| premiumize.me               | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| put.io                      | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| QingStor                    | ModTimeNotSupported                | OK but max 5GiB                     |         | -     | Supports ModTime and SetModTime, but Precision returns ModTimeNotSupported                                        |
| Seafile                     | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| SFTP                        | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Sia                         | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| SugarSync                   | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Tardigrade                  | Nanosecond                         | ErrorCantSetModTime                 |         | R     |                                                                                                                   |
| Uptobox                     | ModTimeNotSupported                | ErrorCantSetModTime                 | Now     | -     | Implementation of ModTime just returns time.Now()                                                                 |
| WebDAV                      | From API                           | ErrorCantSetModTime                 |         | R     | Supports modtimes when used with Owncloud and Nextcloud only                                                      |
| Yandex Disk                 | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Zoho WorkDrive              | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Local                       | Probe disk                         | OK                                  |         | R/W   |                                                                                                                   |

</details>

ncw commented:
> Pushed a new version of the docs, and below is an update to the more detailed table. More to dig into here, but this is as far as I got today.. :)

:-)

I double checked everything in your very nice table and I got this:

- Enterprise File Fabric: the precision will be 1 second for all up to date installations (this was a bug found while making the backend!)
- FTP: After @ivandeex latest changes the FTP backend does support modtimes for sensible FTP servers
- HTTP: arguable. Should probably be `-` since you can't upload stuff, but maybe `R` makes more sense!

albertony commented:
Quick update based on your points on Enterprise File Fabric and FTP. Kept HTTP as is for now.

<details>
<summary>Updated table: Click to expand!</summary>

| Name                        | Precision                          | SetModTime                          | ModTime | Result| Comment                                                                                                           |
| --------------------------- |----------------------------------- | ----------------------------------- | ------- | ----- | ----------------------------------------------------------------------------------------------------------------- |
| 1Fichier                    | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Amazon Drive                | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Amazon S3                   | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Backblaze B2                | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Box                         | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Citrix ShareFile            | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Dropbox                     | Second                             | ErrorCantSetModTimeWithoutDelete    |         | R     |                                                                                                                   |
| Enterprise File Fabric      | Second                             | OK                                  |         | R/W   | Second precision on version 2006.02 and later, hour for older versions                                            |
| FTP                         | Second                             | OK                                  |         | R/W   | Depending on FTP server capabilities, full support on major FTP servers                                           |
| Google Cloud Storage        | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Google Drive                | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Google Photos               | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| HDFS                        | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| HTTP                        | Second                             | "error: http remotes are read only" |         | R     | Precision is the remote http file system's modtime precision, which we have no way of knowing. We estimate at 1s. |
| Hubic                       | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Jottacloud                  | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Koofr                       | ModTimeNotSupported or ms (config) | ErrorCantSetModTimeWithoutDelete    |         | -     |                                                                                                                   |
| Mail.ru Cloud               | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Mega                        | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Memory                      | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Microsoft Azure Blob Storage| Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| OneDrive                    | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| OpenDrive                   | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| OpenStack Swift             | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| pCloud                      | Second                             | ErrorCantSetModTime                 |         | R     |                                                                                                                   |
| premiumize.me               | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| put.io                      | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| QingStor                    | ModTimeNotSupported                | OK but max 5GiB                     |         | -     | Supports ModTime and SetModTime, but Precision returns ModTimeNotSupported                                        |
| Seafile                     | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| SFTP                        | Second                             | OK                                  |         | R/W   |                                                                                                                   |
| Sia                         | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| SugarSync                   | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Tardigrade                  | Nanosecond                         | ErrorCantSetModTime                 |         | R     |                                                                                                                   |
| Uptobox                     | ModTimeNotSupported                | ErrorCantSetModTime                 | Now     | -     | Implementation of ModTime just returns time.Now()                                                                 |
| WebDAV                      | From API                           | ErrorCantSetModTime                 |         | R     | Supports modtimes when used with Owncloud and Nextcloud only                                                      |
| Yandex Disk                 | Nanosecond                         | OK                                  |         | R/W   |                                                                                                                   |
| Zoho WorkDrive              | ModTimeNotSupported                | ErrorCantSetModTime                 |         | -     |                                                                                                                   |
| Local                       | Probe disk                         | OK                                  |         | R/W   |                                                                                                                   |

</details>

ncw commented:
I forgot to merge this - bah!

Can you rebase and I'll merge - thank you :-)

albertony commented:
Rebased and fixed conflicts, i.e. updated with changes from master:
- Akamai Netstorage: New backend, ModTime=R/W
- Internet Archive: New backend, ModTime=R/W with note that option `wait_archive` must be set to non-zero value (if not then Precision returns ModTimeNotSupported). (The option is mentioned in backend docs, but its effect on modtime is not explained).
- Tardigrade renamed to Storj

ncw reviewed (approved):
Looks great - thank you!

Please squash and merge ASAP before we add any more backends!

Note `s/clearify/clarify/` in commit message!
